babel/babel
# Refactor @babel/types to be 100% ES-module

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          | n
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | n
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

TL;DR:
 I want to
* Make @babel/types a real ES-module
* Generate helpers at buildtime instead of runtime
* Split some functions into separate files to avoid dependency cycles (`validators.js` into `validators/isScope.js`, `validators/validate.js` ... ).
* Exports of index.js should stay the same.
* USe the chance and activate flow in all files.

Long version:

Okay I discovered a little rabbit hole here and started going down. Let me quickly explain what the intention is:

I looked at `babel-standalone` and noticed that we still do the commonjs transform there with babel-loader. This makes the webpack `ModuleConcatenationPlugin` pretty much useless. Disabling the commonjs transform does not change anything as babel doesn't get build as ES-module but only as commonjs in `lib/`.
To fix this i though it would be nice to also publish all babel packages with a `module` field in `package.json` and generate a version without commonjs transform.

The problem is that some packages still use commonjs in the source, `@babel/types` being one of them. In there we dynamically generate all the validators and builders like `isIdentifier(node)` `assertIdentifier(node)` and `identifier(name)`. 
Dynamically generating exports does not work in esmodules, so instead I switched to pregenerate all this helpers at buildtime instead of doing it at runtime. This increases the size of the package, which doesn't matter for nodejs and for web builds it should then be able to pick only the exports that are used and tree shake the rest away.

The rest of the changes is trying to get rid of dependency cycles, and to avoid them it would be best that most functions, which are right now in grouped files like `validators.js` should be in separate files.

~~The only one breaking change is that with the buildtime approach we cannot have the builders `import()` and `super()` because they are reserved keywords. (`Import()` and `Super()` still work.)~~
Just figured out that this can be done by using `export { buildImport as import };`


What do you think about this approach? Would that be something that we want? Some other ideas?
